File: src/table.py
"""
table.py — 64×64 Cayley table generator for Kamea Scheme.

Generates the 4KB table that implements:
  - Algebraic core (retraction pair, classifier, composition)
  - Type-dispatched operations (car/cdr/cons on typed values)
  - Immediate fixnum arithmetic (0-31)
  - Type predicates via classifier

The table is the runtime. Everything else is scaffolding.
"""

# ═══════════════════════════════════════════════════════════════════
# Element indices
# ═══════════════════════════════════════════════════════════════════

# Algebraic core
TOP   = 0   # ⊤, nil, empty list
BOT   = 1   # ⊥, #f, error
Q     = 2   # quote / encoder
E     = 3   # eval / decoder
CAR   = 4   # first projection
CDR   = 5   # second projection
CONS  = 6   # pair constructor

# Control
RHO   = 7   # branch / conditional
APPLY = 8   # explicit apply
CC    = 9   # call/cc marker

# Classifier and Y combinator
TAU   = 10  # classifier / type predicate
Y     = 11  # Y combinator (fixed point)
ETA   = Y   # alias — CDR (5) is the composite element, not ETA

# Type tags
T_PAIR = 12
T_SYM  = 13
T_CHAR = 14
T_VEC  = 15
T_STR  = 16
T_CLS  = 17

# Immediate fixnums: 18..49 represent integers 0..31
FIXNUM_BASE = 18
FIXNUM_COUNT = 32

def fixnum(n):
    """Element index for immediate fixnum n (0 <= n <= 31)."""
    assert 0 <= n <= 31, f"fixnum out of range: {n}"
    return FIXNUM_BASE + n

def fixnum_val(elem):
    """Integer value of a fixnum element, or None if not a fixnum."""
    if FIXNUM_BASE <= elem < FIXNUM_BASE + FIXNUM_COUNT:
        return elem - FIXNUM_BASE
    return None

# Built-in symbols
SYM_LAMBDA = 50
SYM_IF     = 51
SYM_DEFINE = 52
SYM_QUOTE  = 53
SYM_SET    = 54
SYM_BEGIN  = 55
SYM_COND   = 56
SYM_ELSE   = 57

# Special values
TRUE  = 58  # #t
EOF   = 59  # eof-object
VOID  = 60  # void / unspecified
TAIL  = 61  # tail-call marker

# Reserved
RESERVED_0 = 62
RESERVED_1 = 63

N = 64  # table size

# Element names for display
NAMES = ['⊤', '⊥', 'Q', 'E', 'car', 'cdr', 'cons',
         'ρ', 'apply', 'cc',
         'τ', 'Y',
         'T_pair', 'T_sym', 'T_char', 'T_vec', 'T_str', 'T_cls'] + \
        [str(i) for i in range(32)] + \
        ['λ', 'if', 'define', 'quote', 'set!', 'begin', 'cond', 'else'] + \
        ['#t', 'eof', 'void', 'tail', '_62', '_63']


# ═══════════════════════════════════════════════════════════════════
# Table construction
# ═══════════════════════════════════════════════════════════════════

def make_table():
    """Build the 64×64 Cayley table."""
    T = [[BOT] * N for _ in range(N)]

    # --- Absorber rows ---
    # ⊤ · x = ⊤ for all x  (top absorber = nil)
    for j in range(N):
        T[TOP][j] = TOP
    # ⊥ · x = ⊥ for all x  (bottom absorber = #f)
    for j in range(N):
        T[BOT][j] = BOT

    # --- Retraction pair (Q/E) ---
    # E(Q(x)) = x for all non-absorber x
    # Q(E(x)) = x for all non-absorber x
    # Q(⊤) = some fixed encoding, Q(⊥) = some fixed encoding
    # E(⊤) = ⊤ (E-transparency), E(⊥) = ⊥ (E-transparency)
    for j in range(N):
        if j == TOP:
            T[E][j] = TOP   # E-transparency
            T[Q][j] = fixnum(0)  # Q(⊤) = encoding of ⊤
        elif j == BOT:
            T[E][j] = BOT   # E-transparency
            T[Q][j] = fixnum(1)  # Q(⊥) = encoding of ⊥
        else:
            # For core elements, Q encodes and E decodes
            # Full specification TBD during table search
            pass

    # --- Classifier (τ) ---
    # τ(x) returns the type tag of x
    T[TAU][TOP] = TOP        # τ(nil) = nil (it's nil)
    T[TAU][BOT] = BOT        # τ(#f) = #f
    T[TAU][TRUE] = TRUE      # τ(#t) = #t (boolean)
    # Type tags for typed values
    for j in range(FIXNUM_BASE, FIXNUM_BASE + FIXNUM_COUNT):
        T[TAU][j] = fixnum(0)  # τ(fixnum) = "number" marker
    # τ(type-tag) = type-tag (type tags classify themselves)
    for j in [T_PAIR, T_SYM, T_CHAR, T_VEC, T_STR, T_CLS]:
        T[TAU][j] = j

    # --- Fixnum arithmetic ---
    # This is the big payoff: add(a, b) as table lookup for small values.
    # We define element-level addition for the fixnum block.
    # T[fixnum(a)][fixnum(b)] = fixnum(a+b) if a+b <= 31, else overflow marker
    for a in range(32):
        for b in range(32):
            s = a + b
            if s <= 31:
                T[fixnum(a)][fixnum(b)] = fixnum(s)
            else:
                T[fixnum(a)][fixnum(b)] = BOT  # overflow → runtime handles it

    # --- Type dispatch for car/cdr ---
    # car applied to a type tag determines behavior
    # (actual pair car/cdr goes through heap; the table handles type checking)
    T[CAR][T_PAIR] = T_PAIR   # car of a pair → valid (runtime resolves)
    T[CAR][T_SYM]  = BOT      # car of a symbol → error
    T[CAR][T_CHAR] = BOT      # car of a char → error
    T[CAR][T_VEC]  = BOT      # car of a vector → error
    T[CAR][T_STR]  = BOT      # car of a string → error
    T[CAR][T_CLS]  = BOT      # car of a closure → error
    for j in range(FIXNUM_BASE, FIXNUM_BASE + FIXNUM_COUNT):
        T[CAR][j] = BOT       # car of a number → error

    T[CDR][T_PAIR] = T_PAIR   # cdr of a pair → valid
    T[CDR][T_SYM]  = BOT
    T[CDR][T_CHAR] = BOT
    T[CDR][T_VEC]  = BOT
    T[CDR][T_STR]  = BOT
    T[CDR][T_CLS]  = BOT
    for j in range(FIXNUM_BASE, FIXNUM_BASE + FIXNUM_COUNT):
        T[CDR][j] = BOT

    # TODO: Fill remaining rows to satisfy:
    #   - Extensionality (all rows distinct)
    #   - ICP (internal composition property)
    #   - Branch semantics
    #   - Composition (η)
    #   - Y combinator (if included)
    #   - Full type dispatch for all operators × all types

    return T


def print_table(T):
    """Print the table in a readable format."""
    print(f"{'':>6s}", end='')
    for j in range(N):
        print(f"{NAMES[j]:>5s}", end='')
    print()
    for i in range(N):
        print(f"{NAMES[i]:>6s}", end='')
        for j in range(N):
            print(f"{NAMES[T[i][j]]:>5s}", end='')
        print()


def emit_c_table(T):
    """Emit the table as a C const array."""
    print("static const uint8_t cayley[64][64] = {")
    for i in range(N):
        row = ', '.join(f"{T[i][j]:2d}" for j in range(N))
        print(f"    {{{row}}},  /* {NAMES[i]} */")
    print("};")


def emit_stats(T):
    """Print table statistics."""
    # Check extensionality
    rows = [tuple(T[i]) for i in range(N)]
    unique = len(set(rows))
    print(f"Rows: {N}, Unique: {unique}, "
          f"{'EXTENSIONAL' if unique == N else f'COLLISIONS: {N - unique}'}")

    # Check absorbers
    top_ok = all(T[TOP][j] == TOP for j in range(N))
    bot_ok = all(T[BOT][j] == BOT for j in range(N))
    print(f"⊤ absorber: {'OK' if top_ok else 'FAIL'}")
    print(f"⊥ absorber: {'OK' if bot_ok else 'FAIL'}")

    # Fixnum arithmetic check
    correct = 0
    total = 0
    for a in range(32):
        for b in range(32):
            total += 1
            s = a + b
            if s <= 31:
                if T[fixnum(a)][fixnum(b)] == fixnum(s):
                    correct += 1
            else:
                if T[fixnum(a)][fixnum(b)] == BOT:
                    correct += 1
    print(f"Fixnum addition: {correct}/{total} correct")


if __name__ == '__main__':
    T = make_table()
    emit_stats(T)
    print()
    print("C table output:")
    emit_c_table(T)

File: src/table.rs
//! The 32×32 Cayley table (1KB).
//!
//! Every type dispatch decision is a single table lookup.
//! The table lives in `.rodata` (flash on embedded).

pub const N: usize = 32;

// ── Algebraic core (0-11) ────────────────────────────────────────

pub const TOP: u8 = 0;     // ⊤, nil, '()
pub const BOT: u8 = 1;     // ⊥, #f, type error
pub const Q: u8 = 2;       // quote / encoder
pub const E: u8 = 3;       // eval / decoder
pub const CAR: u8 = 4;     // first projection
pub const CDR: u8 = 5;     // second projection (= ρ ∘ cons)
pub const CONS: u8 = 6;    // pair constructor, core-preserving
pub const RHO: u8 = 7;     // branch (conditional dispatch)
pub const APPLY: u8 = 8;   // explicit application
pub const CC: u8 = 9;      // call/cc
pub const TAU: u8 = 10;    // classifier (type predicate)
pub const Y: u8 = 11;      // Y combinator (fixed point)

// ── R4RS type tags (12-19) ───────────────────────────────────────
// These are rib tag values. A rib with tag T_PAIR is a pair, etc.

pub const T_PAIR: u8 = 12;
pub const T_SYM: u8 = 13;
pub const T_CLS: u8 = 14;  // closure / procedure
pub const T_STR: u8 = 15;
pub const T_VEC: u8 = 16;
pub const T_CHAR: u8 = 17;
pub const T_CONT: u8 = 18; // first-class continuation
pub const T_PORT: u8 = 19;

// ── Special values (20-22) ───────────────────────────────────────

pub const TRUE: u8 = 20;   // #t
pub const EOF: u8 = 21;    // eof-object
pub const VOID: u8 = 22;   // void (unspecified)

// ── The table ────────────────────────────────────────────────────

pub static CAYLEY: [[u8; N]; N] = [
    [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],  // ⊤
    [ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1],  // ⊥
    [ 2,  2,  3,  2, 11,  4,  6,  7,  9,  5,  8, 10,  0,  2,  2, 13,  2,  2,  2,  2,  2,  2,  2,  2,  0,  2,  2,  2,  2,  2,  2,  2],  // Q
    [ 0,  1,  3,  2,  5,  9,  6,  7, 10,  8, 11,  4,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],  // E
    [ 2,  0,  1,  3,  2,  0, 10,  6,  0, 10,  0, 23, 12,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],  // car
    [ 1,  2,  2, 10,  6, 23, 10,  0, 10,  6,  0,  6, 12,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  0,  0,  0,  0,  0,  0,  0,  0,  1],  // cdr
    [ 0,  2,  4,  9,  8, 11,  9, 10,  6,  8, 10,  7,  2,  2,  2,  2,  2,  2,  2,  2,  2,  0,  2,  0,  0,  0,  0,  3,  0,  2,  2,  2],  // cons
    [ 2,  2,  1,  3,  2, 11, 10,  6,  6, 10,  0, 23,  2,  2,  2,  2,  2, 31,  2,  2,  2,  2,  2,  2,  2,  2,  0,  0,  0,  0,  0, 11],  // ρ
    [ 0,  2,  4,  4, 11, 11,  7,  2,  2,  2,  0, 11,  3,  2,  2,  2,  2,  0,  2,  2,  2,  2,  2,  2,  0,  2,  2,  1,  2,  0,  2,  3],  // apply
    [ 0,  3,  4,  3,  6,  2,  2,  2, 10,  9, 11, 11,  4,  2,  2,  2,  0,  2,  3,  2,  2,  2,  2,  0,  2,  2,  2,  0,  0,  0,  2,  4],  // cc
    [ 0,  1,  0,  0,  0,  1,  0,  0,  1,  0,  0,  0, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22,  0,  1,  0,  0,  0,  0,  0,  0,  0],  // τ
    [ 2,  2,  2,  3,  9, 11,  7,  3, 10,  9,  2, 11,  0,  2,  2,  2,  2,  2,  2,  2,  2,  2, 15,  2,  2,  2,  0,  1,  3,  2,  2,  5],  // Y
    [ 0,  1, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12],  // pair
    [ 0,  1, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13],  // sym
    [ 0,  1, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14],  // cls
    [ 0,  1, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15],  // str
    [ 0,  1, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16],  // vec
    [ 0,  1, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17, 17],  // char
    [ 0,  1, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18, 18],  // cont
    [ 0,  1, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19, 19],  // port
    [ 0,  1, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20],  // #t
    [ 0,  1, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21],  // eof
    [ 0,  1, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22, 22],  // void
    [ 0, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23],
    [ 0, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24, 24],
    [ 0, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25, 25],
    [ 0, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26],
    [ 0, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27, 27],
    [ 0, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28],
    [ 0, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29, 29],
    [ 0, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30, 30],
    [ 0, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31],
];

/// Apply element `a` to element `b` via the Cayley table.
#[inline(always)]
pub fn dot(a: u8, b: u8) -> u8 {
    CAYLEY[a as usize][b as usize]
}

#[cfg(test)]
mod tests {
    use super::*;

    const CORE: [u8; 10] = [Q, E, CAR, CDR, CONS, RHO, APPLY, CC, TAU, Y];
    const TYPE_TAGS: [u8; 8] = [T_PAIR, T_SYM, T_CLS, T_STR, T_VEC, T_CHAR, T_CONT, T_PORT];

    #[test]
    fn absorbers() {
        for j in 0..N as u8 {
            assert_eq!(dot(TOP, j), TOP);
            assert_eq!(dot(BOT, j), BOT);
        }
    }

    #[test]
    fn extensionality() {
        let mut rows: Vec<Vec<u8>> = (0..N).map(|i| CAYLEY[i].to_vec()).collect();
        rows.sort();
        rows.dedup();
        assert_eq!(rows.len(), N);
    }

    #[test]
    fn retraction_qe() {
        for &x in &CORE {
            assert_eq!(dot(E, dot(Q, x)), x, "E(Q({x}))≠{x}");
            assert_eq!(dot(Q, dot(E, x)), x, "Q(E({x}))≠{x}");
        }
    }

    #[test]
    fn e_transparency() {
        assert_eq!(dot(E, TOP), TOP);
        assert_eq!(dot(E, BOT), BOT);
    }

    #[test]
    fn classifier_core() {
        for &x in &CORE {
            let tx = dot(TAU, x);
            assert!(tx == TOP || tx == BOT, "τ({x})={tx}");
        }
    }

    #[test]
    fn classifier_type_tags() {
        for &t in &TYPE_TAGS {
            assert_eq!(dot(TAU, t), t, "τ should return type tag for type {t}");
        }
    }

    #[test]
    fn branch() {
        for &x in &CORE {
            if dot(TAU, x) == TOP {
                assert_eq!(dot(RHO, x), dot(CAR, x));
            } else {
                assert_eq!(dot(RHO, x), dot(CONS, x));
            }
        }
    }

    #[test]
    fn composition() {
        for &x in &CORE {
            assert_eq!(dot(CDR, x), dot(RHO, dot(CONS, x)));
        }
    }

    #[test]
    fn cons_core_preserving() {
        for &x in &CORE {
            let gx = dot(CONS, x);
            assert!(gx >= 2 && gx < 12, "cons({x})={gx} not in core");
        }
    }

    #[test]
    fn y_fixed_point() {
        let yp = dot(Y, RHO);
        assert!(yp >= 2);
        assert_eq!(dot(RHO, yp), yp);
    }

    #[test]
    fn car_type_dispatch() {
        assert_eq!(dot(CAR, T_PAIR), T_PAIR);
        for &t in &TYPE_TAGS {
            if t != T_PAIR {
                assert_eq!(dot(CAR, t), BOT, "car({t}) should be error");
            }
        }
    }
}
